### ISSUE TITLE
"<T>() → T can't be assigned to () → void" through getter

### ISSUE THREAD

**Issue by nex3** [dart-lang/sdk#33015]
The following code:

```dart
void main() {
  dontCall(neverCalled);
}

void dontCall(void function()) {}

T Function<T>() get neverCalled => _neverCalled;

T _neverCalled<T>() => throw "don't call me";
```

Produces this error when run in the VM with `--preview-dart-2`:

```
test.dart:2:12: Error: A value of type '<T extends dart.core::Object>() → T' can't be assigned to a 
variable of type '() → void'.                                                                      
Try changing the type of the left hand side, or casting the right hand side to '() → void'.
  dontCall(neverCalled);
           ^
```

This seems incorrect; `T` should be able to be instantiated to `void`, and in fact the same code works if `_neverCalled` is passed directly to `dontCall()` rather than going through a getter.

This makes it very difficult to use the [`neverCalled`](https://www.dartdocs.org/documentation/test/latest/test/neverCalled.html) getter in `test` in Dart 2 code.

**Comment by eernstg** [comment]
**tl;dr** It's working as intended (for now) **rd;lt**

There is an [upcoming feature specification](https://dart-review.googlesource.com/c/sdk/+/47043) on generic function instantiation&mdash;that is, the feature that you're relying on in order to get `neverCalled` transformed into a non-generic function when it is passed to `dontCall`&mdash;and the example is actually working as it should according to that feature specification.

The reason why "it doesn't work" (that is, `neverCalled` doesn't get transformed into something that we could express as `neverCalled<void>`, if we had had that syntax, so the program has a compile-time error) is that this feature applies only to statically resolved functions and instance methods, it cannot be applied to first class functions.

So when you use `_neverCalled` you're directly denoting a function declaration, and hence generic function instantiation can take place; but when it goes through the getter `neverCalled` you're attempting to perform the generic instantiation for an already existing function object, and this is not supported.

This exact restriction is being debated. I've asked for reasons for having the restriction (which has been around for a long time), and basically nobody can confirm that it is necessary to have it. Still, we might yet receive feedback confirming that it would indeed be very costly (in terms of implementation time, run-time performance, or whatever) to eliminate it.

In other words, it's working as intended relative to the current specification proposal, and even though it might be required to work in the future, there is no guarantee.
